examples/corgi_adventure/application.py

```diff
@@ -0,0 +1,142 @@
+             {RESTRICTIONS}. You respond ONLY with a json object containing two fields: "solved", which is a
+             boolean indicating whether the challenge was solved by the attempt, and "what_happened",
+             which is a string containing a brief narrative, written in the second person and addressed
+             to the player, of what happened during the protagonist's attempt""",
+            },
+            {
+                "role": "user",
+                "content": f"The current challenge is: {state['current_challenge']} "
+                f"and the player's attempt is: {state['attempts'][-1]}",
+            },
+        ],
+    )
+    content = result.choices[0].message.content
+    try:
+        json_result = json.loads(content)
+    except json.JSONDecodeError as e:
+        print("bad json: ", content)
+        json_result = {"solved": False, "what_happened": "Not sure, really. I'm a dog. I can't read json. I can't read at all."}
+
+    result = {"challenge_solved": json_result["solved"], "txt_result": content}
+
+    return result, state.update(
+        challenge_solved=result["challenge_solved"],
+        what_happened=json_result["what_happened"],
+    )
+
+
+@action(
+    reads=["challenge_solved", "current_challenge", "what_happened"],
+    writes=["current_challenge", "did_win"],
+)
+def maybe_progress(state: State) -> Tuple[dict, State]:
+    print("What happened:", state["what_happened"])
+    if state["challenge_solved"]:
+        if state["current_challenge"] == challenges[-1]:
+            result = {"did_win": True}
+        else:
+            result = {
+                "current_challenge": challenges[
+                    challenges.index(state["current_challenge"]) + 1
+                ]
+            }
+    else:
+        result = {"current_challenge": state["current_challenge"]}
+    return result, state.update(**result)
+
+
+@action(reads=["challenges"], writes=[])
+def win(state: State) -> Tuple[dict, State]:
+    # get summary of actions taken from openai
+    print("you won")
+    return {}, state
+
+
+def application(
+    app_id: Optional[str] = None,
+    storage_dir: Optional[str] = "~/.burr",
+    hooks: Optional[List[LifecycleAdapter]] = None,
+) -> Application:
+    return (
+        burr.core.ApplicationBuilder()
+        .with_state(did_win=False)
+        .with_actions(
+            start=start,
+            prompt_for_challenge=prompt_for_challenge,
+            evaluate_attempt=evaluate_attempt,
+            maybe_progress=maybe_progress,
+            win=win,
+        )
+        .with_transitions(
+            ("start", "prompt_for_challenge", default),
+            ("prompt_for_challenge", "evaluate_attempt", default),
+            ("evaluate_attempt", "maybe_progress", default),
+            ("maybe_progress", "win", when(did_win=True)),
+            ("maybe_progress", "prompt_for_challenge", default),
+        )
+        .with_entrypoint("start")
+        .with_tracker(
```

Comment:
Call it `demo:dog_adventure` (case less important than `demo:`) -- this fits in with the rest and will soon display nicely in the UI when rn